libavutil/hwcontext_mediafoundation.c

```diff
@@ -0,0 +1,445 @@
+/*
+ * This file is part of FFmpeg.
+ *
+ * FFmpeg is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * FFmpeg is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with FFmpeg; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+#define COBJMACROS
+#if !defined(_WIN32_WINNT) || _WIN32_WINNT < 0x0602
+#undef _WIN32_WINNT
+#define _WIN32_WINNT 0x0602
+#endif
+
+#pragma comment(lib, "mfuuid.lib")
+
+#include <windows.h>
+#include <initguid.h>
+#include <wmcodecdsp.h>
+#include <mftransform.h>
+#include <mfapi.h>
+#include <mferror.h>
+#include <mfidl.h>
+#include <uuids.h>
+#include <codecapi.h>
+#include <d3d11.h>
+#include <d3d9.h>
+#include <dxva2api.h>
+
+#include <stdint.h>
+#include <string.h>
+
+#include "avassert.h"
+#include "buffer.h"
+#include "common.h"
+#include "imgutils.h"
+#include "hwcontext.h"
+#include "hwcontext_mediafoundation.h"
+#include "hwcontext_internal.h"
+#include "mem.h"
+#include "pixfmt.h"
+#include "pixdesc.h"
+
+typedef struct MFDeviceContext {
+    HANDLE d3d11_dll;
+    HANDLE d3d9_dll;
+    HANDLE dxva2_dll;
+    IDirect3D9 *d3d9;
+} MFDeviceContext;
+
+typedef IDirect3D9* WINAPI pDirect3DCreate9(UINT);
+typedef HRESULT WINAPI pCreateDeviceManager9(UINT *, IDirect3DDeviceManager9 **);
+
+static void mf_uninit_d3d(AVHWDeviceContext *ctx)
+{
+    AVMFDeviceContext *hwctx = ctx->hwctx;
+    MFDeviceContext *priv = ctx->internal->priv;
+
+    if (hwctx->d3d11_manager)
+        IMFDXGIDeviceManager_Release(hwctx->d3d11_manager);
+    hwctx->d3d11_manager = NULL;
+
+    if (priv->d3d11_dll)
+        FreeLibrary(priv->d3d11_dll);
+    priv->d3d11_dll = NULL;
+
+    if (hwctx->d3d9_manager)
+        IDirect3DDeviceManager9_Release(hwctx->d3d9_manager);
+    hwctx->d3d9_manager = NULL;
+
+    if (priv->d3d9)
+        IDirect3D9_Release(priv->d3d9);
+    priv->d3d9 = NULL;
+
+    if (priv->d3d9_dll)
+        FreeLibrary(priv->d3d9_dll);
+    priv->d3d9_dll = NULL;
+
+    if (priv->dxva2_dll)
+        FreeLibrary(priv->dxva2_dll);
+    priv->dxva2_dll = NULL;
+}
+
+static int mf_create_d3d11_device(AVHWDeviceContext *ctx, int loglevel)
+{
+    AVMFDeviceContext *hwctx = ctx->hwctx;
+    MFDeviceContext *priv = ctx->internal->priv;
+    HRESULT hr;
+    UINT token;
+    ID3D11Device *d3d11_device = NULL;
+    ID3D10Multithread *multithread;
```

Comment:
It is unclear
https://learn.microsoft.com/en-us/windows/win32/api/d3d11_4/nn-d3d11_4-id3d11multithread
But could we not stick all in D3D11?
ID3D10Multithread -> ID3D11Multithread

Reply:
Thank you for catching this, yes I agree.  This was another leftover from the PW patch.  I will get it upgraded.